### PR TITLE
[Postponed] Implement synchronous executor (fixes #11)

### LINKED ISSUE #11: Add synchronous Executor
We currently have a sequential executor, but it's not the same as running the code synchronously. For example, in the case of an error, the next job is executed, instead of failing immediately.

I think it would be helpful to have an executor which executes its function directly on `submit()` and returns a resolved future.

### PR DESCRIPTION
**Pull request by philippotto**

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -125,4 +125,24 @@ def execute_with_log_level(log_level):
...
+        raise Exception("Error")
+
+
+    try:
+        with cluster_tools.get_executor("synchronous") as executor:
+            fut = executor.submit(mark_synchronous_execution)
+            assert(was_executed)
+
+            fut = executor.submit(throw)
+        assert(False)
```
Comment (georgwiese) [inline test.py]: Just curious, any reason why you use parentheses for the asserts?

File: cluster_tools/__init__.py [hunk 2]
```diff
@@ -330,10 +330,41 @@ def __init__(self, **kwargs):
...
 
+class SynchronousExecutor():
+
+    def map(self, func, args, timeout=None, chunksize=None):
+        if timeout is not None or chunksize is not None:
+            logging.warning("Timeout and chunksize parameters are currently ignored within the SynchronousExecutor.")
+
+        futs = []
+        for arg in args:
+            self.submit(fun, arg)
```
Comment (georgwiese) [inline cluster_tools/__init__.py]: Append the future?

File: cluster_tools/__init__.py [hunk 3]
```diff
@@ -330,10 +330,41 @@ def __init__(self, **kwargs):
...
+            result = fun(arg)
+            fut.set_result(result)
+        except Exception as e:
+            fut.set_exception(e)
+
+        # This line will raise exceptions immediately. This inverses
+        # the standard handling of futures which can contain errors without
+        # raising them. Now, the call side has to wrap the map call in an additional
+        # try-catch.
+        fut.result()
```
[the remark this answers is not in the document]
Comment (georgwiese) [reply cluster_tools/__init__.py]: How about not catching it in the first place? Then, the debugger would stop it the right place.

File: test.py [hunk 4]
```diff
@@ -125,4 +125,24 @@ def execute_with_log_level(log_level):
...
     debug_out = execute_with_log_level(logging.INFO)
-    assert(not (test_output_str in debug_out))
\ No newline at end of file
+    assert(not (test_output_str in debug_out))
+
+
+def test_synchronous_executor():
+    was_executed = False
+    def mark_synchronous_execution():
+        return was_executed
```
Comment (georgwiese) [inline test.py]: how does this work? Shouldn't it set `was_executed` to true?

**Comment by philippotto** [comment]
As per our discussion, we will postpone this PR until it becomes relevant again.

**Pull request closed without merging**